working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1682,130 +1887,138 @@ At runtime, a pattern is matched against a value. This determines whether or not
 
-    1.  If `v` is not a subtype of `p` then the match fails. *This is a
-        deliberate failure when using a typed variable pattern in a switch in
-        order to test a value's type. In a binder, this can only occur on a
-        failed downcast from `dynamic` and becomes a runtime exception.*
+    1.  If `v` is not a subtype of `p` then the match fails.
 
     2.  Otherwise, bind the variable's identifier to `v` and the match succeeds.
 
-*   **Cast binder**:
+*   **Cast**:
 
     1.  If `v` is not a subtype of `p` then throw a runtime exception. *Note
         that we throw even if this appears in a refutable context. The intent
         of this pattern is to assert that a value *must* have some type.*
 
-    2.  Otherwise, bind the variable's identifier to `v`. The match always
-        succeeds (if it didn't throw).
+    2.  Otherwise, bind the variable's identifier to `v` and the match succeeds.
 
-*   **Null-assert binder**:
+*   **Grouping**: Match the subpattern against `v` and succeed if it matches.
 
-    1.  If `v` is null then throw a runtime exception. *Note that we throw even
-        if this appears in a refutable context. The intent of this pattern is to
-        assert that a value *must* not be null.*
+*   **List**:
 
-    2.  Otherwise, match the inner pattern against `v`.
+    1.  If `v` does not implement `List<T>` for some `T`, then the match fails.
 
-*   **Literal matcher** or **constant matcher**: The pattern matches if `o == v`
-    evaluates to `true` where `o` is the pattern's value.
+    2.  If the length of the list determined by calling `length` is not equal to
+        the number of subpatterns, then the match fails.
 
-    **TODO: Should this be `v == o`?**
+    3.  Otherwise, for each element subpattern, in source order:
 
-*   **Wildcard binder or matcher**: Always succeeds.
+        1.  Extract the element value `e` by calling `[]` on `v` with an
+            appropriate integer index.
 
-*   **Extractor matcher**:
+        2.  Match `e` against the element subpattern.
 
-    1.  If `v` is not a subtype of the extractor pattern's type, then the
-        match fails.
+    4.  The match succeeds if all subpatterns match.
+
+*   **Map**:
+
+    1.  If the value's type does not implement `Map<K,V>` for some `K` and `V`,
+        then the match fails.
+
+    2.  Otherwise, for each entry in `p`:
+
+        1.  Evaluate the key `expression` to `k` and call `containsKey()` on the
+            value. If this returns `false`, the map does not match.
+
+        3.  Otherwise, evaluate `v[k]` and match the resulting value against
+            this entry's value subpattern. If it does not match, the map does
+            not match.
+
+    3.  The match succeeds if all entry subpatterns match.
+
+    *Note that, unlike with lists, a matched map may have additional entries
+    that are not checked by the pattern.*
+
+*   **Record**:
+
+    1.  If `v` is not a record with the same type as `p`, then the match fails.
```

Comment:
Probably "same shape", not "type".

Reply:
Ah, good catch. In the initial version of this, yes, the record type was too precise and could fail unnecessarily.

I've tweaked the rules for calculating the record pattern's static type to mirror the matched value's type, so I think it's safe to use "same type" here. (And it avoids needing to define some runtime notion of checking a record's shape.)

The only failure mode is when the matched value isn't a record type. For that, when calculating the record pattern's type, *there* I say that we synthesize a record type with the same shape as the pattern and `Object?` for all of the fields. That way, the record pattern itself only checks the shape, and then the field subpatterns take it from there.